File: sdks/node/src/dashboard/auth/oauth/flow.ts
// End-to-end OAuth flow orchestration: the seam between the HTTP layer and
// the providers. Owns the provider registry, the state store, and the
// AuthStore integration.

import type { Queue } from "../../../index";
import { createLogger } from "../../../utils";
import { isSafeRedirect } from "../../urlSafety";
import { AuthStore, type DashboardSession } from "../store";
import { callbackUrl, configuredProviders, type OAuthConfig } from "./config";
import {
  IdentityFetchError,
  type OAuthProvider,
  ProviderNotConfigured,
  StateValidationError,
} from "./identity";
import { s256Challenge } from "./pkce";
import type { FetchLike } from "./providers";
import { GenericOidcProvider, GitHubProvider, GoogleProvider } from "./providers";
import { OAuthStateStore } from "./stateStore";

const log = createLogger("dashboard");

/** Instantiate one provider per configured slot, keyed by slot. */
export function buildProviders(
  config: OAuthConfig,
  http: FetchLike = fetch,
): Map<string, OAuthProvider> {
  const registry = new Map<string, OAuthProvider>();
  for (const entry of configuredProviders(config)) {
    if (entry.type === "google") {
      registry.set(entry.slot, new GoogleProvider(entry, http));
    } else if (entry.type === "github") {
      registry.set(entry.slot, new GitHubProvider(entry, http));
    } else {
      registry.set(entry.slot, new GenericOidcProvider(entry, http));
    }
  }
  return registry;
}

/** Ties together config, providers, state store, and the auth store. */
export class OAuthFlow {
  private readonly providers: Map<string, OAuthProvider>;
  private readonly stateStore: OAuthStateStore;

  constructor(
    private readonly queue: Queue,
    private readonly config: OAuthConfig,
    options: { providers?: Map<string, OAuthProvider>; stateStore?: OAuthStateStore } = {},
  ) {
    this.providers = options.providers ?? buildProviders(config);
    this.stateStore = options.stateStore ?? new OAuthStateStore(queue);
    if (this.providers.size > 0 && config.adminEmails.length === 0) {
      // OAuth users only ever get the viewer role without an allowlist, so an
      // OAuth-only deployment would silently have zero admins.
      log.warn(
        () =>
          "OAuth is configured without admin emails: every OAuth login gets the " +
          "viewer role. Set TASKITO_DASHBOARD_OAUTH_ADMIN_EMAILS (or " +
          "OAuthConfig.adminEmails) to grant admin access.",
      );
    }
  }

  get passwordAuthEnabled(): boolean {
    return this.config.passwordAuthEnabled;
  }

  hasProvider(slot: string): boolean {
    return this.providers.has(slot);
  }

  /** Compact provider summary for the login UI (no secrets). */
  providersListing(): Array<{ slot: string; label: string; type: string }> {
    return [...this.providers.values()].map((p) => ({
      slot: p.slot,
      label: p.label,
      type: p.type,
    }));
  }

  /**
   * Mint a state row and return the provider's authorize URL. `nextUrl` is
   * sanitised against {@link isSafeRedirect}, falling back to `/`.
   */
  async start(slot: string, nextUrl: string | null): Promise<string> {
    const provider = this.requireProvider(slot);
    const safeNext = nextUrl && isSafeRedirect(nextUrl) ? nextUrl : "/";
    const state = this.stateStore.create(slot, safeNext);
    return provider.authorizationUrl({
      state: state.state,
      nonce: state.nonce,
      codeChallenge: s256Challenge(state.codeVerifier),
      redirectUri: callbackUrl(this.config, slot),
    });
  }

  /**
   * Exchange `code` for an identity and create a session. Returns the
   * session plus the post-login redirect target. Throws
   * {@link StateValidationError} / {@link IdentityFetchError} /
   * {@link AllowlistDenied} for the handler layer to map.
   */
  async handleCallback(
    slot: string,
    params: { code: string | null; stateToken: string | null; error: string | null },
  ): Promise<{ session: DashboardSession; nextUrl: string }> {
    if (params.error) {
      throw new IdentityFetchError(`provider returned error: ${params.error}`);
    }
    if (!params.code || !params.stateToken) {
      throw new StateValidationError("missing code or state parameter");
    }

    const row = this.stateStore.consume(params.stateToken);
    if (!row) {
      throw new StateValidationError("state is invalid, expired, or already used");
    }
    if (row.slot !== slot) {
      throw new StateValidationError("state slot does not match callback slot");
    }

    const provider = this.requireProvider(slot);
    const identity = await provider.exchangeCode({
      code: params.code,
      codeVerifier: row.codeVerifier,
      redirectUri: callbackUrl(this.config, slot),
      expectedNonce: row.nonce,
    });
    provider.checkAllowlist(identity);

    const store = new AuthStore(this.queue);
    const user = store.getOrCreateOauthUser({
      slot: identity.slot,
      subject: identity.subject,
      email: identity.email,
      name: identity.name,
      emailVerified: identity.emailVerified,
      adminEmails: this.config.adminEmails,
    });
    return { session: store.createSession(user), nextUrl: row.nextUrl };
  }

  /** Best-effort sweep of expired state rows. */
  pruneState(): number {
    return this.stateStore.pruneExpired();
  }

  private requireProvider(slot: string): OAuthProvider {
    const provider = this.providers.get(slot);
    if (!provider) {
      throw new ProviderNotConfigured(`OAuth provider '${slot}' is not configured`);
    }
    return provider;
  }
}

File: sdks/python/taskito/dashboard/oauth/flow.py
"""End-to-end OAuth flow orchestration.

:class:`OAuthFlow` is the seam between the HTTP handler layer and the
provider implementations. It owns the registry of configured providers,
the state store, and the :class:`AuthStore` integration. Handlers call
``start()`` to mint a redirect URL and ``handle_callback()`` to land a
session.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from taskito.dashboard.auth import AuthStore
from taskito.dashboard.oauth.config import (
    GitHubConfig,
    GoogleConfig,
    OAuthConfig,
    OIDCConfig,
)
from taskito.dashboard.oauth.identity import (
    IdentityFetchError,
    OAuthProvider,
    ProviderNotConfigured,
    StateValidationError,
)
from taskito.dashboard.oauth.pkce import s256_challenge
from taskito.dashboard.oauth.providers import (
    GenericOIDCProvider,
    GitHubProvider,
    GoogleProvider,
)
from taskito.dashboard.oauth.state_store import OAuthStateStore
from taskito.dashboard.url_safety import is_safe_redirect

if TYPE_CHECKING:
    from taskito.app import Queue
    from taskito.dashboard.auth import Session

logger = logging.getLogger("taskito.dashboard.oauth")


def build_providers(
    config: OAuthConfig,
) -> dict[str, OAuthProvider]:
    """Instantiate one provider per configured slot, keyed by slot."""
    registry: dict[str, OAuthProvider] = {}
    for entry in config.providers():
        if isinstance(entry, GoogleConfig):
            registry[entry.slot] = GoogleProvider(entry)
        elif isinstance(entry, GitHubConfig):
            registry[entry.slot] = GitHubProvider(entry)
        elif isinstance(entry, OIDCConfig):
            registry[entry.slot] = GenericOIDCProvider(entry)
    return registry


class OAuthFlow:
    """Ties together config, providers, state store, and the auth store."""

    def __init__(
        self,
        queue: Queue,
        config: OAuthConfig,
        *,
        providers: dict[str, OAuthProvider] | None = None,
        state_store: OAuthStateStore | None = None,
    ) -> None:
        self._queue = queue
        self._config = config
        self._providers: dict[str, OAuthProvider] = (
            providers if providers is not None else build_providers(config)
        )
        self._state_store = state_store or OAuthStateStore(queue)
        if self._providers and not config.admin_emails:
            # OAuth users only ever get the viewer role without an allowlist,
            # so an OAuth-only deployment would silently have zero admins.
            logger.warning(
                "OAuth is configured without admin emails: every OAuth login "
                "gets the viewer role. Set TASKITO_DASHBOARD_OAUTH_ADMIN_EMAILS "
                "(or OAuthConfig.admin_emails) to grant admin access."
            )

    # ── Introspection ────────────────────────────────────────────────

    @property
    def password_auth_enabled(self) -> bool:
        return self._config.password_auth_enabled

    def has_provider(self, slot: str) -> bool:
        return slot in self._providers

    def providers_listing(self) -> list[dict[str, str]]:
        """Compact provider summary for the login UI (no secrets)."""
        return [
            {"slot": p.slot, "label": p.label, "type": p.type} for p in self._providers.values()
        ]

    # ── Flow ─────────────────────────────────────────────────────────

    def start(self, slot: str, next_url: str | None) -> str:
        """Mint a state row and return the provider's authorize URL.

        ``next_url`` is sanitised against :func:`is_safe_redirect` and falls
        back to ``"/"`` if it fails the check.
        """
        provider = self._require_provider(slot)
        safe_next = next_url if next_url and is_safe_redirect(next_url) else "/"
        state = self._state_store.create(slot=slot, next_url=safe_next)
        challenge = s256_challenge(state.code_verifier)
        return provider.authorization_url(
            state=state.state,
            nonce=state.nonce,
            code_challenge=challenge,
            redirect_uri=self._config.callback_url(slot),
        )

    def handle_callback(
        self,
        slot: str,
        *,
        code: str | None,
        state_token: str | None,
        error: str | None,
    ) -> tuple[Session, str]:
        """Exchange ``code`` for an identity and create a session.

        Returns ``(session, next_url)`` on success. Raises:

        - :class:`StateValidationError` for missing/expired/replayed state
        - :class:`IdentityFetchError` for any token / userinfo / claim issue
        - :class:`AllowlistDenied` if the identity is outside the allowlist
        """
        if error:
            raise IdentityFetchError(f"provider returned error: {error}")
        if not code or not state_token:
            raise StateValidationError("missing code or state parameter")

        row = self._state_store.consume(state_token)
        if row is None:
            raise StateValidationError("state is invalid, expired, or already used")
        if row.slot != slot:
            raise StateValidationError("state slot does not match callback slot")

        provider = self._require_provider(slot)
        identity = provider.exchange_code(
            code=code,
            code_verifier=row.code_verifier,
            redirect_uri=self._config.callback_url(slot),
            expected_nonce=row.nonce,
        )
        provider.check_allowlist(identity)

        store = AuthStore(self._queue)
        user = store.get_or_create_oauth_user(
            slot=identity.slot,
            subject=identity.subject,
            email=identity.email,
            name=identity.name,
            email_verified=identity.email_verified,
            admin_emails=self._config.admin_emails,
        )
        session = store.create_session(user)
        return session, row.next_url

    # ── Maintenance ──────────────────────────────────────────────────

    def prune_state(self) -> int:
        return self._state_store.prune_expired()

    # ── Internal ─────────────────────────────────────────────────────

    def _require_provider(self, slot: str) -> OAuthProvider:
        provider = self._providers.get(slot)
        if provider is None:
            raise ProviderNotConfigured(f"OAuth provider {slot!r} is not configured")
        return provider
